File: mobile/apps/photos_tv/android/app/src/main/kotlin/io/ente/photos_tv/SlideshowService.kt
package io.ente.photos_tv

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request
import kotlin.math.max
import kotlin.random.Random

internal class SlideshowService(
    private val client: OkHttpClient,
    private val payload: CastPayload,
    private val imageCache: DiskImageCache,
    private val cryptoBox: CryptoBox,
) {
    private var files = emptyList<CastFile>()
    private var index = 0

    fun nextImage(): ByteArray? {
        if (index >= files.size) refreshFiles()
        if (files.isEmpty()) return null
        val file = files[index]
        index += 1
        return downloadImage(file)
    }

    private fun refreshFiles() {
        val decryptedFiles = mutableListOf<CastFile>()
        for (remoteFile in getRemoteFiles()) {
            val file = decryptFile(remoteFile)
            if (file != null && file.isImage) decryptedFiles.add(file)
        }
        files = decryptedFiles.shuffled(Random.Default)
        index = 0
    }

    private fun getRemoteFiles(): List<JsonObject> {
        val filesById = linkedMapOf<Long, JsonObject>()
        var sinceTime = 0L
        while (true) {
            val request = Request.Builder()
                .url("$API_ORIGIN/cast/diff?sinceTime=$sinceTime")
                .header("X-Cast-Access-Token", payload.castToken)
                .get()
                .build()
            client.newCall(request).execute().use { response ->
                response.ensureOk()
                val body = JsonConfig.value.parseToJsonElement(response.body!!.string()).jsonObject
                for (item in body.getValue("diff").jsonArray.map { it.jsonObject }) {
                    sinceTime = max(sinceTime, item.getLong("updationTime"))
                    if (item["isDeleted"]?.jsonPrimitive?.booleanOrNull == true) filesById.remove(item.getLong("id")) else filesById[item.getLong("id")] = item
                }
                if (body["hasMore"]?.jsonPrimitive?.booleanOrNull != true) return filesById.values.toList()
            }
        }
    }

    private fun decryptFile(item: JsonObject): CastFile? {
        val key = cryptoBox.decrypt(
            input = cryptoBox.base64Decode(item.getString("encryptedKey")),
            key = cryptoBox.base64Decode(payload.collectionKey),
            nonce = cryptoBox.base64Decode(item.getString("keyDecryptionNonce")),
        )
        val metadata = item.getValue("metadata").jsonObject
        val metadataBytes = cryptoBox.decryptData(
            input = cryptoBox.base64Decode(metadata.getString("encryptedData")),
            key = key,
            header = cryptoBox.base64Decode(metadata.getString("decryptionHeader")),
        )
        return castFileFromRemote(item, key, metadataBytes)
    }

    private fun downloadImage(file: CastFile): ByteArray {
        val cachedImage = imageCache.read(file)
        if (cachedImage != null) {
            imageCache.markUsed(file)
            return cachedImage
        }
        val imageBytes = downloadAndDecryptImage(file)
        imageCache.write(file, imageBytes)
        return imageBytes
    }

    private fun downloadAndDecryptImage(file: CastFile): ByteArray {
        val request = Request.Builder()
            .url("$CAST_WORKER_ORIGIN/preview/?fileID=${file.id}")
            .header("X-Cast-Access-Token", payload.castToken)
            .get()
            .build()
        client.newCall(request).execute().use { response ->
            response.ensureOk()
            return cryptoBox.decryptData(
                input = response.body!!.bytes(),
                key = file.key,
                header = cryptoBox.base64Decode(file.preview.getString("decryptionHeader")),
            )
        }
    }
}

File: mobile/apps/photos_tv/lib/src/slideshow_service.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:ente_crypto_api/ente_crypto_api.dart';
import 'package:http/http.dart' as http;

import 'config.dart';
import 'http_util.dart';
import 'models.dart';

/// Service that fetches and decrypts slideshow images.
class SlideshowService {
  final http.Client _client;
  final CastPayload _payload;
  final _random = Random();
  var _files = <CastFile>[];
  var _index = 0;

  /// Creates slideshow service.
  SlideshowService(this._client, this._payload);

  /// Returns next slideshow image bytes.
  Future<Uint8List?> nextImage() async {
    if (_index >= _files.length) await _refreshFiles();
    if (_files.isEmpty) return null;
    final file = _files[_index];
    _index += 1;
    return _downloadImage(file);
  }

  /// Closes HTTP resources.
  void close() => _client.close();

  Future<void> _refreshFiles() async {
    final remoteFiles = await _getRemoteFiles();
    final files = <CastFile>[];
    for (final remoteFile in remoteFiles) {
      final file = await _decryptFile(remoteFile);
      if (file != null && file.isImage) files.add(file);
    }
    _files = files;
    _files.shuffle(_random);
    _index = 0;
  }

  Future<List<Map<String, dynamic>>> _getRemoteFiles() async {
    final filesByID = <int, Map<String, dynamic>>{};
    var sinceTime = 0;
    while (true) {
      final response = await _client.get(
        Uri.parse('$apiOrigin/cast/diff?sinceTime=$sinceTime'),
        headers: {'X-Cast-Access-Token': _payload.castToken},
      );
      ensureOk(response);
      final body = jsonDecode(response.body) as Map<String, dynamic>;
      final diff = body['diff'] as List<dynamic>;
      for (final item in diff.cast<Map<String, dynamic>>()) {
        sinceTime = max(sinceTime, item['updationTime'] as int);
        if (item['isDeleted'] == true) {
          filesByID.remove(item['id'] as int);
        } else {
          filesByID[item['id'] as int] = item;
        }
      }
      if (body['hasMore'] != true) return filesByID.values.toList();
    }
  }

  Future<CastFile?> _decryptFile(Map<String, dynamic> item) async {
    final key = CryptoUtil.decryptSync(
      CryptoUtil.base642bin(item['encryptedKey'] as String),
      CryptoUtil.base642bin(_payload.collectionKey),
      CryptoUtil.base642bin(item['keyDecryptionNonce'] as String),
    );
    final metadata = item['metadata'] as Map<String, dynamic>;
    final metadataBytes = await CryptoUtil.decryptData(
      CryptoUtil.base642bin(metadata['encryptedData'] as String),
      key,
      CryptoUtil.base642bin(metadata['decryptionHeader'] as String),
    );
    return CastFile.fromRemote(item, key, metadataBytes);
  }

  Future<Uint8List> _downloadImage(CastFile file) async {
    final response = await _client.get(
      Uri.parse('$castWorkerOrigin/preview/?fileID=${file.id}'),
      headers: {'X-Cast-Access-Token': _payload.castToken},
    );
    ensureOk(response);
    final preview = file.preview;
    return CryptoUtil.decryptData(
      response.bodyBytes,
      file.key,
      CryptoUtil.base642bin(preview['decryptionHeader'] as String),
    );
  }
}
